fim_process_unit_wb.sh

```diff
@@ -0,0 +1,119 @@
+    echo 'NOTE: fim_pre_processing must have been already run and this tool'
+    echo '      will not include post processing. Only single independent single'
+    echo '      huc and its branches.'    
+    echo 'Usage : There are no arg keys (aka.. no dashes)'
+    echo '        you need the run name first, then the huc.'
+    echo '        ie ) /foss_fim/fim_process_unit_wb.sh rob_test_1 05030104'
+    echo
+    exit
+}
+
+export runName=$1
+export hucNumber=$2
+
+# print usage if arguments empty
+if [ "$runName" = "" ]
+then
+    echo "ERROR: Missing run time name argument (1st argument)"
+    usage
+fi
+
+if [ "$hucNumber" = "" ]
+then
+    echo "ERROR: Missing hucNumber argument (2nd argument)"
+    usage
+fi
+
+re='^[0-9]+$'
+if ! [[ $hucNumber =~ $re ]] ; then
+   echo "Error: hucNumber is not a number" >&2; exit 1
+   usage
+fi
+
+echo "=========================================================================="
+echo "---- Start of huc processing for $hucNumber"
+
+# outputDataDir, srcDir and others come from the Dockerfile
+
+export outputRunDataDir=/$outputDataDir/$runName
+export outputHucDataDir=$outputRunDataDir/$hucNumber
+export outputBranchDataDir=$outputHucDataDir/branches
+export current_branch_id=0
+
+## huc data
+if [ -d "$outputHucDataDir" ]; then
+    rm -rf $outputHucDataDir
+fi
+
+# make outputs directory
+mkdir -p $outputHucDataDir
+mkdir -p $outputBranchDataDir
+
+# Clean out previous unit logs and branch logs starting with this huc
+rm -f $outputRunDataDir/logs/unit/"$hucNumber"_unit.log
+rm -f $outputRunDataDir/logs/branch/"$hucNumber"_summary_branch.log
+rm -f $outputRunDataDir/logs/branch/"$hucNumber"*.log
+rm -f $outputRunDataDir/unit_errors/"$hucNumber"*.log
+rm -f $outputRunDataDir/branch_errors/"$hucNumber"*.log
+hucLogFileName=$outputRunDataDir/logs/unit/"$hucNumber"_unit.log
+
+# Process the actual huc
+/usr/bin/time -v $srcDir/run_unit_wb.sh | tee $hucLogFileName
+
+#exit ${PIPESTATUS[0]} (and yes.. there can be more than one)
+return_codes=( "${PIPESTATUS[@]}" )
+
+#echo "huc return codes are:"
+#echo $return_codes
+
+# we do this way instead of working directly with stderr and stdout
+# as they were messing with output logs which we always want.
+for code in "${return_codes[@]}"
+do
+    # Make an extra copy of the branch log in a new folder  if an error
+    # Note: It was tricky to load in the fim_enum into bash, so we will just 
+    # go with the code for now
+    if [ $code -eq 0 ]; then
+        echo
+        # do nothing
+    elif [ $code -eq 60 ]; then
+        echo
+        echo "***** Unit has no valid branches *****"
+    elif [ $code -eq 61 ]; then
+        echo
+        echo "***** Unit has not a valid unit *****"        
+    else
+        echo
+        echo "***** An error has occured  *****"
+        # copy the error log over to the unit_errors folder to better isolate it
+        cp $hucLogFileName $outputRunDataDir/unit_errors
```

Comment:
Should we include this line in the two `elif` blocks above as well, so the error is captured and displayed in the `/unit_errors` logs. 

Here is a path to what a log file currently looks like for a bad huc run through `fim_pipeline.sh`: 
`/foss_fim/outputs/test_robhpr806_fimpreprocessing_badhuc/unit_errors/043001009_unit.log`

Reply:
Yes. I want to add that but decided to skip it for now in favor of letting post processing handle it. I need to clean the unit_errors in certainn conditions and am worried about letting the huc do it and how it handles the non_zero_exit_codes.log against multi core

Reply:
Sounds great.